### PR TITLE
Only request stats of columns needed for cardinality estimation [#150…

### PR DESCRIPTION
**Pull request by oarap**
…424379]

GPORCA should not spend time extracting column statistics that are not
needed for cardinality estimation. This PR eliminates this overhead
of requesting and generating the statistics for columns that are not
used in cardinality estimation unnecessarily.

E.g:
`CREATE TABLE foo (a int, b int, c int);`

For table foo, the query below only needs for stats for column `a` which
is the distribution column and column `c` which is the column used in
where clause.
`select * from foo where c=2;`

However, prior to that commit, the column statistics for column `b` is
also calculated and passed for the cardinality estimation. The only
information needed by the optimizer is the `width` of column `b`. For
this tiny information, we transfer every stats information for that
column.

This PR and its counterpart PR in GPORCA [https://github.com/greenplum-db/gporca/pull/228](url) ensures that the column width information is passed and extracted in the `dxl:Relation` metadata
information.

Preliminary results for short running queries provides up to 65x
performance improvement.

Signed-off-by: Jemish Patel <jpatel@pivotal.io>

### CODE REVIEW

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 1]
```diff
@@ -751,6 +751,22 @@ CTranslatorRelcacheToDXL::Pdrgpmdcol
...
+		{
+			DOUBLE dWidth = CStatistics::DDefaultColumnWidth.DVal();
+			if (ulColLen == ULONG_MAX)
+				ulColLen = (ULONG) dWidth;
+		}
+		else
+		{
+			Form_pg_statistic fpsStats = (Form_pg_statistic) GETSTRUCT(heaptupleStats);
+
+			// column width
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: can u add a comment on why u want to overwrite this?

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 2]
```diff
@@ -751,6 +751,22 @@ CTranslatorRelcacheToDXL::Pdrgpmdcol
 			ulColLen = (ULONG) att->atttypmod - VARHDRSZ;
 		}
 
+		// extract out histogram and mcv information from pg_statistic
+		HeapTuple heaptupleStats = gpdb::HtAttrStats(rel->rd_id, ul+1);
+		if (!HeapTupleIsValid(heaptupleStats))
+		{
+			DOUBLE dWidth = CStatistics::DDefaultColumnWidth.DVal();
+			if (ulColLen == ULONG_MAX)
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: can u add a comment on why u want to overwrite this?

File: src/include/gpopt/translate/CTranslatorUtils.h [hunk 3]
```diff
@@ -126,6 +126,10 @@ namespace gpdxl
 			static
 			const CWStringConst *PstrSystemColName(AttrNumber attno);
 
+			// returns the length for the sytem column with given number
```
Comment (vraghavan78) [inline src/include/gpopt/translate/CTranslatorUtils.h]: system

File: src/backend/gpopt/translate/CTranslatorUtils.cpp [hunk 4]
```diff
@@ -876,6 +903,49 @@ CTranslatorUtils::PmdidSystemColType
 	}
 }
 
+
+// Returns the length for the sytem column with given number
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorUtils.cpp]: system

Comment (vraghavan78) [reply src/backend/gpopt/translate/CTranslatorUtils.cpp]: for a given attno

File: src/include/gpopt/translate/CTranslatorUtils.h (hunk 3 above)
Comment (vraghavan78) [reply src/include/gpopt/translate/CTranslatorUtils.h]: for a given attno

File: src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp [hunk 5]
```diff
@@ -751,6 +751,22 @@ CTranslatorRelcacheToDXL::Pdrgpmdcol
...
+				ulColLen = (ULONG) dWidth;
+		}
+		else
+		{
+			Form_pg_statistic fpsStats = (Form_pg_statistic) GETSTRUCT(heaptupleStats);
+
+			// column width
+			ulColLen = fpsStats->stawidth;
+		}
+
```
Comment (dhanashreek89) [inline src/backend/gpopt/translate/CTranslatorRelcacheToDXL.cpp]: `heaptupleStats` should be freed using `gpdb::FreeHeapTuple()`

File: src/backend/gpopt/translate/CTranslatorUtils.cpp [hunk 6]
```diff
@@ -170,6 +172,31 @@ CTranslatorUtils::Pdxltabdesc
...
+			}
+			else
+			{
+				Form_pg_statistic fpsStats = (Form_pg_statistic) GETSTRUCT(heaptupleStats);
+
+				// column width
+				ulWidth = fpsStats->stawidth;
+			}
+		}
+		pmdtype->Release();
```
Comment (dhanashreek89) [inline src/backend/gpopt/translate/CTranslatorUtils.cpp]: Same here. `heaptupleStats` needs to be freed

File: src/test/regress/expected/update_gp_optimizer.out [hunk 7]
```diff
@@ -97,43 +97,44 @@ WHERE t1.user_vie_project_code_pk = keo1.user_vie_project_code_pk;
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Update  (cost=0.00..2586.11 rows=1 width=1)
    ->  Result  (cost=0.00..2586.00 rows=1 width=42)
-         ->  Explicit Redistribute Motion 1:3  (slice7)  (cost=0.00..2586.00 rows=2 width=38)
+         ->  Explicit Redistribute Motion 3:3  (slice7; segments: 3)  (cost=0.00..2586.00 rows=1 width=38)
                ->  Result  (cost=0.00..2586.00 rows=1 width=38)
                      ->  Split  (cost=0.00..2586.00 rows=1 width=38)
                            ->  Result  (cost=0.00..2586.00 rows=1 width=42)
-                                 ->  Hash Join  (cost=0.00..2586.00 rows=1 width=34)
+                                 ->  Hash Join  (cost=0.00..2586.00 rows=1 width=66)
```
Comment (vraghavan78) [inline src/test/regress/expected/update_gp_optimizer.out]: why 66?

**Pull request merged**